### DISCUSSION TITLE
export/import automation to maximize the cost efectiveness

### DISCUSSION THREAD

**Discussion by gustonator** [mletenay/home-assistant-goodwe-inverter#282] [Q&A]
Hi, 
I have an idea, but have absolutely no clue, to to create such automation. (in HA or Node Red - doesn't matter)
What I need:

for Summer: 
1.) look at the spot prices of today / tomorrow
2.) if there are negative spot prices from ie. 13:00 - 18:00, sent every created energy from panels directly to the network, from early morning -  to get best money from it
3.) when the negative spot price at 13:00 occurs (lets say -110eur/MWh), drain energy primary FROM the grid (to charge the battery and for use and to earn money)
4.) if there is only minimal negative spot price (ie. 40 - -40eur/MWh) use energy from panels

and second automation would be 
for the winter time: 
1.) look at the current and next day, get the lowest spot prices, and charge the battery during those times.

Is anybody here who could help?
Maybe there is already some automation for this out there, but I couldn't find it.

many Thanks

**Comment by zhouska** [comment]
Hi,

the basics are quite simple. I'm not claiming it is 100% what you want, but this should give you an idea you can pivot from...

1 + 2. 

Discharge to grid:

I'm only looking at spot price for any given day (no look ahead yet). I have a few other values to check - PV forecast, spot price values to make it worth and dates of interest. You then wait for the spot price to go below certain value and repeat the steps in reverse. It is handy to set up some notifications for the major steps.

![image](https://github.com/mletenay/home-assistant-goodwe-inverter/assets/12584719/6ede8a55-e9d2-4c45-871b-52a5fe4068be)


You need to set the inverted to discharge mode, set 'Eco mode power' to at least 1% and set the value 'Grid export limit' to a value your grid operator allows you to. There is one caveat though, if your PV is not able to sustain household load, it will buy from grid (and not from battery), so you might want to built in some other checks.

Negative spot price:

I'm turning on hot water boilers first (you can tweak it to turn on AC, sauna etc) in order to consume energy in-house and minimize  export to grid without setting the grid export limit switch (inverter uses ~0.1kWh by doing if the limit switch is on). And only when battery reaches 94% do I turn on the grid export limit switch. 

![image](https://github.com/mletenay/home-assistant-goodwe-inverter/assets/12584719/a47b6ab9-cf6e-4958-ae5e-a89fc1b26d97)

3. I don't use it, but you can charge the batteries or heat hot water, but probably not a great idea as your main concern is to consume PV generated energy first (esp. if you don't set the grid export limit switch) 
4. You can achieve it by setting the spot price to some positive value, other than 0

Conversely, same can be done in winter times... 

HTH

**Comment by gustonator** [comment]
Thanks!
In the meantime I started to create my own automation, and now it's in testing phase. 
I also need to add forecast to it.

problem is, that I wanted to prevent spamming the invertor with settings, therefore I'm checking first the current state, and then I set the new one.

I'm also setting the battery export limit/power according spot price and battery charge %. Automation also looks into times, what the user have set up, if it's in the sell times or not... 
I've also implemented a Automatic/manual switch to be able to override the setting... 

the whole schema is quite wild :D

![image](https://github.com/mletenay/home-assistant-goodwe-inverter/assets/43645090/55763fbd-40a4-4496-9e1c-07dcae638a6a)

![image](https://github.com/mletenay/home-assistant-goodwe-inverter/assets/43645090/734821f4-81bf-495a-bc22-4d9748e46807)

**Reply by zhouska** [reply]
> Thanks! In the meantime I started to create my own automation, and now it's in testing phase. I also need to add forecast to it.
> 
> problem is, that I wanted to prevent spamming the invertor with settings, therefore I'm checking first the current state, and then I set the new one.
> 
> I'm also setting the battery export limit/power according spot price and battery charge %. Automation also looks into times, what the user have set up, if it's in the sell times or not... I've also implemented a Automatic/manual switch to be able to override the setting...
> 
> the whole schema is quite wild :D
> 
> ![image](https://private-user-images.githubusercontent.com/43645090/334738847-55763fbd-40a4-4496-9e1c-07dcae638a6a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NzU1ODksIm5iZiI6MTcxNjk3NTI4OSwicGF0aCI6Ii80MzY0NTA5MC8zMzQ3Mzg4NDctNTU3NjNmYmQtNDBhNC00NDk2LTllMWMtMDdkY2FlNjM4YTZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDA5MzQ0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkNzdkNmM5ZmY4OGY2NWYxMDU2NDhiN2U2ODg4ZmZkZDYxMzE2ZGJkMzM5OTA2YWY3M2FkYTRjZmRjOTc5ODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.irB4GS0J3mhNri-3v5L9sJY22yl12ATmT3KyryqHzkk)
> 
> ![image](https://private-user-images.githubusercontent.com/43645090/334738991-734821f4-81bf-495a-bc22-4d9748e46807.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NzU1ODksIm5iZiI6MTcxNjk3NTI4OSwicGF0aCI6Ii80MzY0NTA5MC8zMzQ3Mzg5OTEtNzM0ODIxZjQtODFiZi00OTVhLWJjMjItNGQ5NzQ4ZTQ2ODA3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDA5MzQ0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3ODMwYTI5YjA4MzZkY2Y2ZTAyNDgzYTg0YjY5MDE5OTBhNGY0YTkxNTBlNzlkMzM3MDBkMzZlNTBiN2Q4OTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gEO4ZQ0cwk0TyYyO_7qqEElGobFpFGt5qZ5NrDnDD4I)

oh yeah, it can get complex pretty fast, esp. as you say you don't want to 'bombard' inverter with setting requests and you want to make sure it is robust without many obvious mistakes. 

Btw, I have an elaborate setup about battery level states based on month of the year (winter only), where I take into account PV forecast and set the dampening values based on weather forecast (factoring in cloudy, rainy, snowy weather), to arrive at a desired battery level so I avoid charging it too much....

The only thing I'm missing so far is some conditional loops to allow to factor in how much I consume in-house/send to the grid so I can turn on devices such as hot water, AC, sauna to shave off the load. I think I will have to look into NodeRed for this...

**Reply by gustonator** [reply]
yeah, NR is ideal for this, but it's also complicated in the beginnings to learn it

**Comment by dmshimself** [comment]
I thought I had followed one part of the example, but I'm having some trouble and I wonder if anyone can see where I am going wrong.  My example in my test is firstly to set the eco mode power to 95 and then set the inverter operation mode option to eco_discharge, which is 5.  

When I do this, the battery SOC is at around 100% and I want to stop the discharge to the grid at soc 95%.  After running the two actions shown below, the battery discharges to grid just fine, but keeps going and the battery SOC goes way below 95%.  I've also tried setting eco mode power to 5 in case that is the right number to stop discharging.  A section of my automation looks like this:

![image](https://github.com/user-attachments/assets/b76ac6d0-62f5-44c6-b9c6-4ca891620493)

**Reply by gustonator** [reply]
@dmshimself SoC means State of Charge. That means, that is doesnt affect the discharge. Only the 'Eco Charge' mode. 

tldr: 
if you set SOC to 95, and your battery is ie. at 80%, and you set the Mode to 'Echo charge', your battery will be charged up to the SOC limit - 95%

so... if you want to stop the discharge at some point, you need to implement some IF statetent.:
if battery is less than 95%, set mode to General.

**Comment by gustonator** [comment]
...or you can eventually set the deph of discharge limit. normally it is set in summer to 90 (battery discharges to 10%)
So if you set the discharge limit to 10-20, your battery will be discharged only to 90-80%.

the entity is called:
number.goodwe_depth_of_discharge_on_grid

**Comment by dmshimself** [comment]
Very useful information indeed, much appreciated.  Have you come across any Goodwe docs that describe these?  All the manuals I have found so far are very quiet on subjects like eco_discharge.

**Reply by gustonator** [reply]
Actually I red the documentation, and what wasn't there, I figured out by testing by myself :)
ie: https://pl.goodwe.com/Ftp/EN/Downloads/User%20Manual/GW_ET_User%20Manual-EN.pdf
look for "Modes" or "PV Master" sections. 

also I'm in local GoodWe and HA groups on forums and Facebook where is plenty of information. 
eventually you can try Chat GPT which should also give you some good and actual informations.

**Comment by jan-trnka** [comment]
Hi, it's not exactly what you're looking for, but maybe it could serve as inspiration. I've put together a few blueprints to help optimize PV systems with GoodWe inverters in Home Assistant. Some of them address similar use cases to yours. Here's a link to the GitHub repository with more details: [GitHub repo](https://github.com/jan-trnka/home-assistant-goodwe-pv-control)

P.S. At the moment, the solution is best suited for households that buy electricity at a fixed rate and sell surplus power on the spot market, which might be different from your setup. You’ll also need the Nordpool or Czech Energy Spot Prices integration to enable spot price functionality.

**Comment by FrederikLacroix** [comment]
@gustonator Do you have an English version of the node-red flow published publically somewhere, I am wanting do do something similar and was wondering if there was a way to get the code from your flow to do a deeper analysis on what you are actually doing and keeping into account.
Thanks
